Lab 2 - pre-full-pull

Helbushek opened the pull request:
all BinaryTree methods withoud vertical tree print.

Helbushek committed:
Create Lab-2 folder

Helbushek committed:
Lab-2 template for synchtonizing git and local repositories

Helbushek committed:
unused directory deletion

Helbushek committed:
implemented methids for BinaryTree:
isEmpty
copyChild
height
power
max
min
find and remove are still in progress

Helbushek committed:
Finished search by key (find) function, started on implementing remove function. Have no clue what to do about build files, maybe they will stay, idk

Helbushek committed:
Revert "Finished search by key (find) function, started on implementing remove function. Have no clue what to do about build files, maybe they will stay, idk"

This reverts commit 8763525c8c9574e83c0c614cc77ca529666c5951.

Helbushek committed:
Finished search by key (find) function, started on implementing remove function. Have no clue what to do about build files, maybe they will stay, idk

Helbushek committed:
practice for 21.03 results. Writes ellipse with text inside.

Helbushek committed:
finished implementing all green and yellow methods. Fixed: find.
Finished:
remove
height (by key)
print Lists
checks for Ideal and Balanced
making tree into vector of integers.

Helbushek committed:
Not important, small prepare for qt work

Helbushek committed:
Final commit, implemented drawing nodes on scene, drawing lines between nodes and coloring nodes background to redraw on lines (to not show lines behind nodes). Connected Add and Remove buttons.

Helbushek committed:
Merge pull request #4 from Helbushek/Lab-2-Qt-printable

Final commit, implemented drawing nodes on scene, drawing lines betwe…

Helbushek committed:
Fixed the bug for remove by key function (remove(int key));
This issue WAS a bug, when i thought it is some Qt scene loadings and freezes, but now it is fixed.
Made with a lot of workarounds, like working not with the node that has needed key but with its predecessor to implement deleting connection to leading node if nothing is beyond her. Left comments in code, but I do not recommend to try to get into it, i prefer to accept the way that it is because it works

Helbushek committed:
Merge branch 'Lab-2' of https://github.com/Helbushek/2nd_course_2.0 into Lab-2

tvz-041 commented on Lab-2/BinaryTree.h:
```diff
@@ -0,0 +1,109 @@
+
+    operator std::vector<int>();
+private:
+    BinaryTree(Node* root);
```
```suggestion
    explicit BinaryTree(Node* root);
```
Конструктор несколько спеифичен, лучше избежать неявных срабатываний.

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+   if (root == nullptr || (!root->getRight() && !root->getLeft()))
+       return true;
+   if (qAbs(power(root->getLeft()) - power(root->getRight())) <= 1 && (_isIdeal(root->getLeft()) && _isIdeal(root->getRight())))
+   {
```
Не рекомендую делать основной класс дерева зависимым от Qt, это ухудшит переносимость.

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+        root->setRight(nullptr);
+    }
+}
+void BinaryTree::clearChild(Node *root){
```
```suggestion
void BinaryTree::clearChildren(Node *root){
```

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+
+BinaryTree BinaryTree::copyRightChild() const{
+    return m_root->getRight();
+}
```
Так он не копирует.

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    {
+        return 0;
+    }
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return 1;
+    }
+    if (root->getRight()==nullptr) {
+        return height(root->getLeft())+1;
+    }
+    if (root->getLeft()==nullptr) {
+        return height(root->getRight())+1;
+    }
```
Т.к. обработка нулевого root есть, max ниже отработает как надо. Эти условия по сути не нужны.

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+int BinaryTree::power() const {
+    return power(m_root);
+}
+int BinaryTree::power(Node* root) const {
```
```suggestion
int BinaryTree::degree(Node* root) const {
```

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    {
+        return 0;
+    }
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return 1;
+    }
+    if (root->getRight()==nullptr) {
+        return power(root->getLeft())+1;
+    }
+    if (root->getLeft()==nullptr) {
+        return power(root->getRight())+1;
+    }
```
То же, что с высотой.

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+int BinaryTree::max() const {
+    return max(m_root);
+}
+int BinaryTree::max(Node* root) const {
+    if (root->getRight() == nullptr && root->getLeft() == nullptr) {
+        return root->getKey();
+    }
+    if (root->getRight()==nullptr) {
+        return std::max(max(root->getLeft()), root->getKey());
+    }
+    if (root->getLeft()==nullptr) {
+        return std::max(max(root->getRight()), root->getKey());
+    }
+    return std::max(std::max(max(root->getRight()), max(root->getLeft())), root->getKey());
+}
```
Случай для пустого дерева?

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    return std::min(std::min(min(root->getRight()), min(root->getLeft())), root->getKey());
+}
+
+BinaryTree BinaryTree::find(const int key) const
+{
+    BinaryTree temp = clone(_find(key, m_root));
+    return temp;
+}
```
Почему для поиска нужно клонировать дерево?

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+        return root;
+    }
+    else {
+        Node* searchRight = _find(key, root->getRight());
+        Node* searchLeft = _find(key, root->getLeft());
+        if (searchRight!=nullptr) {
+            return searchRight;
+        }
+        else if (searchLeft!=nullptr) {
+            return searchLeft;
+        }
+        else
+            return nullptr;
```
Зачем сразу искать в левом поддереве, если мы уже могли найти узел в правом?
```suggestion
        Node* node = _find(key, root->getRight());
        if (node == nullptr) {
            node = _find(key, root->getLeft());
        }
        
        return node;
```

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    return false;
+}
+
+BinaryTree::Node* BinaryTree::detect(int key, Node* root) {
+    if (root==nullptr) {
+        return nullptr;
+    }
+    if (
+            (root->getRight()!=nullptr && root->getRight()->getKey()==key) ||
+            (root->getLeft()!=nullptr && root->getLeft()->getKey()==key) )
+    {
+        return root;
+    }
+    else {
+        Node* searchRight = _find(key, root->getRight());
+        Node* searchLeft = _find(key, root->getLeft());
+        if (searchRight!=nullptr) {
+            return searchRight;
+        }
+        else if (searchLeft!=nullptr) {
+            return searchLeft;
+        }
+        else
+            return nullptr;
+    }
+}
```
Что-то в этой функции не так.
Почему в if мы возвращаем родителя узла с ключом, а в else - узел с ключом?

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    }
+}
+
+bool BinaryTree::remove(Node* root) {
+    if (root == nullptr)
+    {
+        return false;
+    }
+    if (!root->hasChilden()) { // if has no child -> return to higher level (or exit, but should not)
+        return true;
+    }
+    else
+    {
+        if (root->getRight()!=nullptr && !root->getRight()->hasChilden()) // if has right child AND right has no children
+        {
+            delete root->getRight(); // delete Right
+            root->setRight(nullptr); // remove right from tree;
+            return false;
+        }
+        else if (root->getRight() != nullptr && root->getRight()->hasChilden()) // if right exist and has child or two, repeat for both his childs
+        {
+            remove(root->getRight());
+
+        }
+
+        if ((root->getLeft() != nullptr && !root->getLeft()->hasChilden()))
+        {
+            delete root->getLeft();
+            root->setLeft(nullptr);
+            return false;
+        }
+        else if (root->getLeft() != nullptr && root->getLeft()->hasChilden()) // if left exist and has child or two, repeat for both his childs
+        {
+            remove(root->getLeft());
+
+        }
+        else
+            return false;
+    }
+}
```
Тут непонятно. Сам root не должен удаляться, что ли?

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    _printList(m_root);
+}
+
+void BinaryTree::_printList(Node *root) const
```
```suggestion
void BinaryTree::_nlrPrint(Node *root) const
```

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    return;
+}
+
+BinaryTree::operator std::vector<int>()
```
```suggestion
BinaryTree::operator std::vector<int>() const
```

tvz-041 commented on Lab-2/BinaryTree.cpp:
```diff
@@ -0,0 +1,488 @@
+    return temp;
+}
+
+// NODE ZONE BELOW \|\|\|
```
В целом, не воспрещается сделать два cpp, привязанных к одному .h, если вам так удобнее. Хотя в этом случае можно и определение класса Node в отдельный .h вынести (правда, инклюд дерева сделать всё равно придётся) - в этом случа можно даже и реализации пометить как nline, поскольку они все очень просты.

tvz-041 commented on Lab-2/TreeNodeGraphicsItem.cpp:
```diff
@@ -0,0 +1,82 @@
+
+TreeNodeGraphicsItem::TreeNodeGraphicsItem(const QString &text, QGraphicsItem *parent)
+	: QGraphicsSimpleTextItem(text, parent)
+	{
```
Отступы здесь и ниже подправить бы.

tvz-041 commented on Lab-2/TreeNodeGraphicsItem.cpp:
```diff
@@ -0,0 +1,82 @@
+    painter->translate(textRect.x(), textRect.y());
+    QGraphicsSimpleTextItem::paint(painter, option, widget);
+}
+
+//QPointF TreeNodeGraphicsItem::pos() const
+//{
+//    return QGraphicsSimpleTextItem::pos();
+//}
+//void TreeNodeGraphicsItem::setPos(qreal X, qreal Y)
+//{
+
+//}
```
```suggestion

```

tvz-041 commented on Lab-2/TreeNodeGraphicsItem.h:
```diff
@@ -0,0 +1,31 @@
+    QColor borderColor();
+    QColor fillColor();
+
+//    QPointF pos() const;
+
+//    void setPos(qreal X, qreal Y);
```
```suggestion
```

Helbushek committed:
Small fixes, and general refactoring.
Fixed (refactored) functions:
_idIdeal
_idBalanced
cloning of tree`s children
degree (previously power for Node*)
max
min
find
_find
detect
remove(Node*)
_nlrPrint (previously _printList)
operator std::vector<int>

Took Node out from BinaryTree file and made separate file for class

Fixed spacing in TreeNodeGraphicsItem.cpp

Added height and power indicators on scene and connected them

Helbushek committed:
final fixes and refactoring. Implementation of testing class.

Helbushek committed:
small refactoring
